internal/services/monitor/monitor_data_collection_rule_resource.go

```diff
@@ -0,0 +1,892 @@
+				Description:       description,
+				Destinations:      destinations,
+				Kind:              kind,
+				Location:          location,
+				Tags:              tag,
+			})
+		},
+		Timeout: 5 * time.Minute,
+	}
+}
+
+func (r DataCollectionRuleResource) Update() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			id, err := datacollectionrules.ParseDataCollectionRuleID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			metadata.Logger.Infof("updating %s..", *id)
+			client := metadata.Client.Monitor.DataCollectionRulesClient
+			resp, err := client.Get(ctx, *id)
+			if err != nil {
+				return fmt.Errorf("retrieving %s: %+v", *id, err)
+			}
+			if resp.Model == nil {
+				return fmt.Errorf("unexpected null model of %s", *id)
+			}
+			existing := resp.Model
+			if existing.Properties == nil {
+				return fmt.Errorf("unexpected null properties of %s", *id)
+			}
+
+			var state DataCollectionRule
+			if err := metadata.Decode(&state); err != nil {
+				return err
+			}
+
+			if metadata.ResourceData.HasChange("kind") {
+				existing.Kind = expandDataCollectionRuleKind(state.Kind)
+			}
+
+			if metadata.ResourceData.HasChange("tags") {
+				existing.Tags = tags.Expand(state.Tags)
+			}
+
+			if metadata.ResourceData.HasChange("data_flow") {
+				existing.Properties.DataFlows = expandDataCollectionRuleDataFlows(state.DataFlows)
+			}
+
+			if metadata.ResourceData.HasChange("data_sources") {
+				dataSource, err := expandDataCollectionRuleDataSources(state.DataSources)
+				if err != nil {
+					return err
+				}
+				existing.Properties.DataSources = dataSource
+			}
+
+			if metadata.ResourceData.HasChange("description") {
+				existing.Properties.Description = utils.String(state.Description)
+			}
+
+			if metadata.ResourceData.HasChange("destinations") {
+				existing.Properties.Destinations = expandDataCollectionRuleDestinations(state.Destinations)
+			}
+
+			// otherwise Service will return an error: "The resource definition is invalid."
+			existing.SystemData = nil
+
+			if _, err := client.Create(ctx, *id, *existing); err != nil {
+				return fmt.Errorf("updating %s: %+v", *id, err)
+			}
+			return nil
+
+		},
+		Timeout: 30 * time.Minute,
+	}
+}
+
+func (r DataCollectionRuleResource) Delete() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.Monitor.DataCollectionRulesClient
+			id, err := datacollectionrules.ParseDataCollectionRuleID(metadata.ResourceData.Id())
+			if err != nil {
+				return err
+			}
+
+			metadata.Logger.Infof("deleting %s..", *id)
+			resp, err := client.Delete(ctx, *id)
+			if err != nil && !response.WasNotFound(resp.HttpResponse) {
+				return fmt.Errorf("deleting %s: %+v", *id, err)
+			}
+			return nil
+		},
+		Timeout: 30 * time.Minute,
+	}
+}
+
+func expandStringTypeSlice[T ~string](input []string) *[]T {
```

Comment:
We're looking at how generics will fit into the provider so we're not quite ready to accept something like this yet. I appreciate the effort put into this but to merge this resource, we'll have to use the current way of mapping types.

Reply:
Thank @mbfrahry for reviewing. Despite the convenience, the use of generics also brings other risks, so it makes sense for not using generics in the provider now. I have changed the code and rerun the tests. Please continue review.
```
--- PASS: TestAccMonitorDataCollectionRule_complete (157.58s)
--- PASS: TestAccMonitorDataCollectionRule_basic (186.86s)
--- PASS: TestAccMonitorDataCollectionRule_requiresImport (222.00s)
--- PASS: TestAccMonitorDataCollectionRule_update (458.63s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/monitor       459.741s

```